docs: add server side authorization documentation

gikf opened the pull request:
- [x] I have read [Chapter's contributing guidelines](https://github.com/freeCodeCamp/chapter/blob/main/CONTRIBUTING.md).
- [x] My pull request has a descriptive title (not a vague title like `Update README.md`).
- [x] My pull request targets the `main` branch of Chapter.

---

- Adds role/permission based authorization documentation.
- As supposedly tables and flowcharts can be easier for comprehension there are both here! :P
- Any comments regarding wording, what might be missing/not needed, or positioning of the section are appreciated.

gikf committed:
docs: add server side authorization documentation

Sboonny commented on CONTRIBUTING.md:
```diff
@@ -431,6 +431,83 @@ After you have added new feature, to make sure it stays working, we recommend us
 
 [This](https://github.com/freeCodeCamp/chapter/contribute) is a good place to go if you are looking to get started.
 
+## Authorizing GraphGQL requests with `@Authorized` decorator
```
I think it's better to be less specific, so it's more comfortable and easier for our future self to broaden it when necessary.

```suggestion
## Authorized GraphQL Requests
```

Sboonny commented on CONTRIBUTING.md:
```diff
@@ -431,6 +431,83 @@ After you have added new feature, to make sure it stays working, we recommend us
 
+## Authorizing GraphGQL requests with `@Authorized` decorator
+
+### Overview or roles
```
```suggestion
### Current Roles
```

Sboonny commented on CONTRIBUTING.md:
```diff
@@ -431,6 +431,83 @@ After you have added new feature, to make sure it stays working, we recommend us
+
+### Overview or roles
+
+Chapter uses role based permission system, with specific permission assigned to one or multiple roles, on one or more different _levels/scopes_).
```
```suggestion
Chapter has a role based permission system, with specific permissions and different **scopes** assigned to each role.
```

Sboonny commented on CONTRIBUTING.md:
```diff
@@ -431,6 +431,83 @@ After you have added new feature, to make sure it stays working, we recommend us
+
+_Note: in MVP Event roles are not used._
+
+### `@Authorized` decorator
```
```suggestion
### `@Authorized` Decorator Usage in Chapter
```

Sboonny commented on CONTRIBUTING.md:
````diff
@@ -431,6 +431,83 @@ After you have added new feature, to make sure it stays working, we recommend us
+import { Authorized } from 'type-graphql';
+```
+
+Adding it in resolver before specific request logic, wraps the request in the authorization logic. Only for authorized users logic of the request will be executed.
````
```suggestion
Adding it in the resolver before specific request logic, will wrap the request in the authorization logic, which will execute only authorized users' logic of the request.
```

gikf replied:
The last part of the suggestion:
> which will execute only authorized users' logic of the request.

Changes the sense of the original sentence. The point is that when user isn't authorized, their request will not reach the actual code intended for handling that request (when user is authorized), but will be stopped earlier, at the authorization step.

Sboonny replied:
What do you think of this?

```suggestion
Structuring your authorization logic, by adding a resolver before the specific request, will wrap that request. With this, the request will execute that logic only for authorized users
```

Sboonny commented on CONTRIBUTING.md:
```diff
@@ -431,6 +431,83 @@ After you have added new feature, to make sure it stays working, we recommend us
+
+Adding it in resolver before specific request logic, wraps the request in the authorization logic. Only for authorized users logic of the request will be executed.
+
+By default this will mean only logged in user can be authorized.
```
```suggestion
By default, this means only logged in users can be authorized.
```

Sboonny commented on CONTRIBUTING.md:
````diff
@@ -431,6 +431,83 @@ After you have added new feature, to make sure it stays working, we recommend us
+  }
+```
+
+To allow authorization logic connecting user roles on specific chapter or event, with their permissions for performing request, certain arguments are required in the request.
````
```suggestion
To allow authorization logic connecting users' roles on a specific chapter or event, with their permissions for performing the request, certain arguments are required in the request.
```

Sboonny commented on CONTRIBUTING.md:
```diff
@@ -431,6 +431,83 @@ After you have added new feature, to make sure it stays working, we recommend us
+
+### Example
+
+Adding `@Authorize(Permissions.EventEdit)` before event editing logic will authorize user if one of the following applies:
```
```suggestion
Adding `@Authorize(Permissions.EventEdit)` before an event -which edits the logic- will authorize the user if one of the following applies:
```

gikf replied:
Could you take another look here? Suggestion is also altering the sense of original sentence.

Sboonny replied:
Sure, that what I originally thought was going one 👍. Hopefully I will think of better suggestion

Sboonny replied:
I had an issue with `event editing logic` because it feels like it's messing something. But that maybe me overthinking, so feel free to discard these suggestions if it's harmful to the overall document

Sboonny commented on CONTRIBUTING.md:
```diff
@@ -431,6 +431,83 @@ After you have added new feature, to make sure it stays working, we recommend us
+* User is not banned on the chapter, on which edited event is, and for that chapter their Chapter role has `Permission.EventEdit` permission.
+* Their Event role, for the edited event, has `Permission.EventEdit` permission.
+
+Otherwise request will be rejected.
```
```suggestion
Otherwise, the request will be rejected.
```

Sboonny reviewed (commented):
Here is how it looks https://github.com/gikf/chapter/blob/docs/role-permission-based-authorization/CONTRIBUTING.md#authorizing-graphgql-requests-with-authorized-decorator 

We can move it to here https://github.com/gikf/chapter/blob/docs/role-permission-based-authorization/CONTRIBUTING.md#api-specification

About wording, here are suggestions

allella commented:
Thanks. It looks like a lot of work went into it, but I can't comment on the technical parts.

I do think it would be good to put the bulk of it in an expandable section, like we have with the "Steps" throughout the docs.

Also, adding a heading and in-page link to the section in the Table of Contents at the top of the document would be good once you determine where it fits in the hierarchy.

gikf committed:
Apply suggestions from code review

Co-authored-by: Muhammed Mustafa <muhammed@freecodecamp.org>

gikf committed:
chore: restructure text

gikf committed:
Merge branch 'docs/role-permission-based-authorization' of github.com:gikf/chapter into docs/role-permission-based-authorization

gikf committed:
fix: wording

Co-authored-by: Muhammed Mustafa <muhammed@freecodecamp.org>

gikf committed:
Merge branch 'main' into docs/role-permission-based-authorization

gikf committed:
fix: correct indentation

gikf committed:
chore: use smaller heading

gikf committed:
fix: leftover parenthesis

gikf commented:
I've moved it to server section. Not being expandable doesn't make it standout here as much, but expandable sections still could be added if needed. I've put it in the TOC and updated as well some wording.

Sboonny reviewed (approved):
LGTM 🎉, thank you for your contribution to the page! 👍.

ojeytonwilliams reviewed (approved):
I'm sorry for letting this sit here for so long, it looks great.

By the way, that flowchart is amazing!  Did you generate it with a tool or just learn the syntax?

gikf commented:
@ojeytonwilliams :rocket: done by hand, with some documentation/examples and live preview.

ojeytonwilliams commented:
Good to know, thanks. I'll have to try that myself,
